### PR TITLE
🏗️🗃️ UnifiedDataLoader rewrite

### PR DESCRIPTION
**Pull request by the-bay-kay**
UnifiedDataLoader is the last remaining service in `/www/services.js` to be rewritten, as discussed in [Issue 977](https://github.com/e-mission/e-mission-docs/issues/977). The remaining services in this file (Chats and ReferHelper) are no longer being used, so my plan is to remove these and then convert `services.js` into `unifiedDataLoader.ts`.  

I did have an open question about the services.js rewrite - once all of the rewritten services are together, (commHelper, controlHelper, and unifiedData), should these be relocated into a directory other than `/www/js/`? I'm not sure whether these would be moved into a new `services/` directory, sorted into the current folders, or kept as they are.  I figured it was worth thinking about how we wanted to organize these going forward, since the `/www/js/` has expanded a bit!

### CODE REVIEW

**Comment by the-bay-kay** [comment]
Just a quick observation about the behavior or `getUnifiedMessagesForInterval`: When called by `/www/js/survey/enketo/enketoHelper.ts`, the return value (serverResponse.phone_data) always seems to return blank.  I.e., the following snippet:
```
var remotePromise = getRawEntries([key], tq.startTs, tq.endTs)
          .then(function(serverResponse: serverData) {
            console.log('-->', JSON.stringify(serverResponse.phone_data));
            return serverResponse.phone_data;
          });
```
Results in the output:
```
[0] [phonegap] [console.log] --> []
```
Should this always be empty?  I've tried this on a few different emulators, timelines, and surveys: all of which had the same result.  If this isn't intentional, I can look into it further - regardless, I'll be committing and pushing my rewrite soon!

**Comment by the-bay-kay** [comment]
From the manual tests I've done, the rewrite seems to be performing well!  I did have two more questions:
- The older code mentions refactoring the two `getUnified` methods into one generalized method.  Is it worth trying to do this, while I'm already re-writing the functions? I'll go ahead and try to rewrite these tomorrow morning.

- Under what context is `getUnifiedSensorDataForInterval()`  called? From what I can tell, it's used in `timelineHelper.ts` for the `getAllUnprocessedInputs()` method, which in turn is called in the `LabelTab` components.  After some time experimenting with different emulators and timelines, I could never get hit the breakpoints I put on these methods.  Any advice on testing these functions would be greatly appreciated!

**Comment by the-bay-kay** [comment]
As discussed in PR #1052, switching the branches cleans the git logs to only contain my edits - I also updated the title, since I accidentally used a duplicate in icons : )

**Comment by the-bay-kay** [comment]
Quick Update: Now that the `unifiedData` functions have been generalized, I'm working on generalizing `combinePromise` so that it can iterate over a list of promises (Like the original To-Do comment suggested).  

My general approach has been to recursively build up the promises.  For example, if we had an array of promises `[p1, p2, p3]`, they would be combined in the order `(p3+p2) + p1`.  Since this function takes two promises as input, and returns a promise, I think this is a relatively intuitive change - that being said, I've had some difficulty formatting the base case.  

I'm pretty sure I'm making some jumps in my logic and missing something obvious - I'll spend some more time on the recursive approach, and switch gears if that doesn't go anywhere.

**Comment by the-bay-kay** [comment]
image
All tests have been passed successfully! Will be testing on Android before marking as Ready for Review.

**Comment by shankari** [comment]
> I did have an open question about the services.js rewrite - once all of the rewritten services are together, (commHelper, controlHelper, and unifiedData), should these be relocated into a directory other than /www/js/? I'm not sure whether these would be moved into a new services/ directory, sorted into the current folders, or kept as they are. I figured it was worth thinking about how we wanted to organize these going forward, since the /www/js/ has expanded a bit!

Good thought! I vote for moving them into a new services/ directory. I think that the directories are within `www/js` (e.g. `www/js/diary`) though

**Comment by shankari** [comment]
> When called by /www/js/survey/enketo/enketoHelper.ts, the return value (serverResponse.phone_data) always seems to return blank. I.e., the following snippet:

What is the key that the `enketoHelper` is trying to read? If it is the demographic result, it is not included in the fake data that we load, so it will always be empty. To check if it is non-empty, you need to:
- save the onboarding survey and "end trip and force sync" OR
- use a stage/production token

If the key is a trip or place confirm or an addition, it may just be that your data has not been pushed to the server. Pick a stage token that is related to time use (I think I have one in the wiki) and test; you should see some results.

Please report the results of this testing before I review further.

**Comment by shankari** [comment]
> Edit: I may be wrong, but I believe these are called during the initial loading and refresh processes! I've also gone ahead and rewritten getUnified, in https://github.com/e-mission/e-mission-phone/commit/36cfff64668d819576c276a97f1c8052f332063f

Initial loading of the label screen, yes.

**Comment by the-bay-kay** [comment]
> Good thought! I vote for moving them into a new services/ directory. I think that the directories are within `www/js` (e.g. `www/js/diary`) though

I went ahead and created a `www/js/services/` directory, and moved the rewrites into there! While I can't view the exact merge conflict that popped up, I'm hoping it should be an easy fix of changing the imports : )  I'll go ahead and move `controlHelper` in my other rewrite into the same directory!

**Comment by shankari** [comment]
@the-bay-kay I just merged in the onboarding changes. You need to resolve merge conflicts

File: www/js/services.js
Comment (shankari) [inline www/js/services.js:223]: Thanks for doing this! FYI, this was part of the original ionic starter project that I used to create the the very first version of the app. I kept it around for a bit as an example, and then nobody ever had the time/chutzpah to clean it up 😄 

I'm glad we are finally getting around to it!

**Review by shankari — CHANGES_REQUESTED** [review]
Looks good at the high level - just a few cleanup changes and a couple of fairly biggish ones where I am not 100% how the tests pass (or maybe I don't understand it myself).

File: www/js/services/unifiedDataLoader.ts
Comment (the-bay-kay) [reply www/js/services/unifiedDataLoader.ts]: I believe you're right, thanks! I'll spend some time tweaking the tests to see why that didn't get caught.

File: www/js/services.js
Comment (JGreenlee) [inline www/js/services.js:274]: There are a bunch of unused images in `www/img` that might be appropriate to also remove at this point in time

Comment (JGreenlee) [reply www/js/services.js:274]: Not just the 'chats' images, but also remnants of Habitica, calorie-counting, and some random gifs.

I think the only image still needed is `enketo_bare_150x56.png`.

**Comment by the-bay-kay** [comment]
Just noticed the failed check - I'm unsure why the`LoadMoreButton` tests are failing.  I'll go ahead and look into that, as the tests are also failing on my `controlHelper` branch.  Here's the error:
image

**Comment by JGreenlee** [comment]
Is this still being worked on? Is it waiting for review?

**Comment by the-bay-kay** [comment]
> Is this still being worked on? Is it waiting for review?

I had been waiting to run the `prettier` setup, but it'll be ready for review after I make one more small adjustment!  I'll go ahead and put it under your 'for-review' shortly : )

**Comment by the-bay-kay** [comment]
At the time of writing, this does not pass the 'prettier' check because of `index.html` - please see PR #1096 for more info!

File: www/js/services/unifiedDataLoader.ts
Comment (JGreenlee) [reply www/js/services/unifiedDataLoader.ts:27]: I don't think it's necessary to do this before merging this PR, but it might be good to keep around as a 'future change'

File: www/js/types/serverData.ts
Comment (JGreenlee) [inline www/js/types/serverData.ts:1]: Great work here - these type defs will surely be helpful in the future

**Review by JGreenlee — CHANGES_REQUESTED** [review]

File: www/js/services.js
Comment (the-bay-kay) [reply www/js/services.js:1]: Oh, oops - totally! I'll go ahead and remove it now.

File: www/js/services/unifiedDataLoader.ts
Comment (the-bay-kay) [reply www/js/services/unifiedDataLoader.ts:27]: I didn't know `Promise.allSettled` was an option, cool! I'll spend a bit of time seeing if I can get that to work : )

File: www/js/services/unifiedDataLoader.ts
Comment (JGreenlee) [inline www/js/services/unifiedDataLoader.ts:9]: Note to self: we do the same deduplication here as we do in https://github.com/e-mission/e-mission-phone/pull/1086/files#diff-00cfb43594908885f4c2e9121a2224b2c080d9b96bff2e09ec6d4f8d8c6b0645R102-R104

Once both are merged it would be nice to use a common function in both places

File: www/js/services/unifiedDataLoader.ts
Comment (JGreenlee) [inline www/js/services/unifiedDataLoader.ts:30]:
```suggestion
        let allRej = true;
        const values = [];
        const rejections = [];
```

Comment (the-bay-kay) [reply www/js/services/unifiedDataLoader.ts:30]: I had no idea a const array could still be changed with its member functions, interesting!

Comment (JGreenlee) [reply www/js/services/unifiedDataLoader.ts]: Yeah `const` just means it can't be reassigned. So you can modify `const` objects all you want as long as that object keeps the same address in memory.

**Review by JGreenlee — APPROVED** [review]
LGTM

**Comment by shankari** [comment]
@the-bay-kay this now has merge conflicts *again*. I am going to try to resolve them to avoid an infinite cycle of changes.

**Comment by shankari** [comment]
Ah, it failed because of prettier and I can't see anything obviously wrong with the file that it is flagging...

**Comment by JGreenlee** [comment]
I would fix it tonight if I could, but I don't have write access to this one

**Comment by JGreenlee** [comment]
@shankari Looks like the only file not passing Prettier is `diaryTypes.ts`. So if you wish to merge this tonight anyway, I can take care of `diaryTypes.ts` tomorrow in https://github.com/e-mission/e-mission-phone/pull/1086 and it will all even out

**Comment by the-bay-kay** [comment]
>Looks like the only file not passing Prettier is `diaryTypes.ts`. So if you wish to merge this tonight anyway, I can take care of `diaryTypes.ts` tomorrow in #1086 and it will all even out

If you'd like, I can fetch the current code, pull from `service_rewrites_2023` to resolve the merge conflicts, and then run prettier on the whole thing before re-pushing the code!

**Comment by shankari** [comment]
@the-bay-kay as you can see, there is a failing test and it is *not* `LoadMoreButton`. Can you please fix before I review?

**Comment by the-bay-kay** [comment]
> @the-bay-kay as you can see, there is a failing test and it is _not_ `LoadMoreButton`. Can you please fix before I review?

Fixed! I missed a couple imports in the merge, all tests are passing now.

**Review by shankari — APPROVED** [review]
This touches 46 files, so I am merging it now. But I expect that this will be cleaned up in your next PR.

**Pull request merged**
